apps/emqx_streams/src/emqx_streams_message_db.erl

```diff
@@ -0,0 +1,699 @@
+        %% So we delete in groups with all different shards.
+        lists:flatmap(
+            fun(Slabs) ->
+                do_delete(DB, Topics, ?STREAMS_MESSAGE_DB_DELETE_RETRY + 1, Slabs, [])
+            end,
+            slabs_by_generation(DB)
+        )
+    end),
+    ?tp_debug(streams_message_db_delete, #{
+        topic => Topics,
+        time => us_to_ms(_Time),
+        errors => Errors
+    }),
+    case Errors of
+        [] ->
+            ok;
+        _ ->
+            {error, Errors}
+    end.
+
+slabs_by_generation(DB) ->
+    ByGeneration = lists:foldl(
+        fun({_Shard, Generation} = Slab, Acc) ->
+            maps:update_with(Generation, fun(Slabs) -> [Slab | Slabs] end, [Slab], Acc)
+        end,
+        #{},
+        maps:keys(emqx_ds:list_slabs(DB))
+    ),
+    maps:values(ByGeneration).
+
+do_delete(_DB, _Topics, 0, _Slabs, Errors) ->
+    Errors;
+do_delete(DB, Topics, Retries, Slabs0, _Errors0) ->
+    Refs = lists:map(
+        fun({Shard, Generation} = Slab) ->
+            TxOpts = #{
+                db => DB,
+                shard => Shard,
+                generation => Generation,
+                sync => false
+            },
+            {async, Ref, _} = emqx_ds:trans(TxOpts, fun() ->
+                lists:foreach(
+                    fun(Topic) ->
+                        emqx_ds:tx_del_topic(Topic)
+                    end,
+                    Topics
+                )
+            end),
+            {Ref, Slab}
+        end,
+        Slabs0
+    ),
+    Errors = lists:filtermap(
+        fun({Ref, Slab}) ->
+            receive
+                ?ds_tx_commit_reply(Ref, Reply) ->
+                    case emqx_ds:tx_commit_outcome(DB, Ref, Reply) of
+                        {ok, _} ->
+                            false;
+                        {error, IsRecoverable, Reason} ->
+                            ?tp(warning, streams_message_db_delete_error, #{
+                                topics => Topics,
+                                is_recoverable => IsRecoverable,
+                                reason => Reason,
+                                slab => Slab
+                            }),
+                            {true, {Slab, {IsRecoverable, Reason}}}
+                    end
+            end
+        end,
+        Refs
+    ),
+    case Errors of
+        [] ->
+            [];
+        _ ->
+            {Slabs, _} = lists:unzip(Errors),
+            timer:sleep(?STREAMS_MESSAGE_DB_DELETE_RETRY_DELAY),
+            do_delete(DB, Topics, Retries - 1, Slabs, Errors)
+    end.
+
+stream_message_topic(#{topic_filter := TopicFilter, id := Id} = _Stream, Key) ->
+    ?STREAMS_MESSAGE_DB_TOPIC(TopicFilter, Id, Key).
+
+db(Stream) ->
+    case emqx_streams_prop:is_append_only(Stream) of
+        true ->
+            ?STREAMS_MESSAGE_REGULAR_DB;
+        false ->
+            ?STREAMS_MESSAGE_LASTVALUE_DB
+    end.
+
+insert_generation(?STREAMS_MESSAGE_LASTVALUE_DB) ->
+    1.
```

Comment:
Nit. Having this indirection makes not much sense, maybe just put it behind a macrodef?